### PR TITLE
challenge table

### LINKED ISSUE #769: Event db: Challenge table.
We need a new Challenge table for the Event Db.

Code for this already exists in https://github.com/0chain/0chain/pull/679 [challenge.go](https://github.com/0chain/0chain/blob/challenge-events/code/go/0chain.net/smartcontract/dbs/event/challenge.go#L10).

This needs to be reworked as part of https://github.com/0chain/0chain/issues/741. The event database code should be altered to use the blobber table rather than the blobberChallenge table.

### PR DESCRIPTION
**Pull request by ssardana08**
## Fixes
- removed PrevID from code

## Changes
- implemented challenge table
- /openchallenges handler and /getchallenge handler using eventsdb

## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test: https://github.com/0chain/system_test/pull/260
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

File: code/go/0chain.net/smartcontract/dbs/event/challenge.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+	if exists {
+		return errors.New("challenge already exists in db. cannot add")
+	}
+
+	result := edb.Store.Get().Create(&ch)
+
+	return result.Error
+}
+
+func (edb *EventDb) updateChallenge(updates dbs.DbUpdates) error {
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/event/challenge.go:69]: we can make it with one query, look
https://stackoverflow.com/a/23824049

File: code/go/0chain.net/smartcontract/dbs/event/challenge.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+	return &ch, nil
+}
+
+func (edb *EventDb) GetOpenChallengesForBlobber(blobberID string, now, cct common.Timestamp) ([]*Challenge, error) {
+	var chs []*Challenge
+	expiry := now - cct
+
+	result := edb.Store.Get().Model(&Challenge{}).
+		Where("blobber_id = ? AND responded = ? AND created_at >= ?",
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/event/challenge.go]: do we have index for this clause?
what cardinality would be for challenge table

File: code/go/0chain.net/smartcontract/storagesc/allocation.go [hunk 3]
```diff
@@ -1534,7 +1534,7 @@ func (sc *StorageSmartContract) finishAllocation(
 			"can't get related challenge pool: "+err.Error())
 	}
 
-	var passPayments float64 = 0.0
+	var passPayments = 0.0
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/storagesc/allocation.go:1537]: ooof, float can hurt, we need to review all smart contracts for float usage

File: code/go/0chain.net/smartcontract/storagesc/challenge_eventdb.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	for _, v := range ch.Validators {
+		validatorsArr = append(validatorsArr, v.ID)
+	}
+	validators := strings.Join(validatorsArr, ",")
+	return &event.Challenge{
+		ChallengeID:    ch.ID,
+		CreatedAt:      ch.Created,
+		AllocationID:   ch.AllocationID,
+		BlobberID:      ch.BlobberID,
+		ValidatorsID:   validators,
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/storagesc/challenge_eventdb.go]: do we ever query "all challenges for current validator" or similar? if yes then we might normalize this field

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | feat/chall_table  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | feat/chall_table  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

File: code/go/0chain.net/smartcontract/storagesc/challenge_eventdb.go (hunk 4 above)
Comment (ssardana08) [reply code/go/0chain.net/smartcontract/storagesc/challenge_eventdb.go]: nope, not as of now

File: code/go/0chain.net/smartcontract/storagesc/allocation.go (hunk 3 above)
Comment (ssardana08) [reply code/go/0chain.net/smartcontract/storagesc/allocation.go:1537]: ahh! @dabasov can you create an issue for the same.
Here I just removed float64 which was not required..

File: code/go/0chain.net/smartcontract/dbs/event/challenge.go (hunk 1 above)
Comment (ssardana08) [reply code/go/0chain.net/smartcontract/dbs/event/challenge.go:69]: fixed it

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | feat/chall_table  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | feat/chall_table  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by ssardana08** [comment]
```                                         Table "public.challenges"
     Column      |           Type           | Collation | Nullable |                Default
-----------------+--------------------------+-----------+----------+----------------------------------------
 id              | bigint                   |           | not null | nextval('challenges_id_seq'::regclass)
 created_at      | bigint                   |           |          |
 updated_at      | timestamp with time zone |           |          |
 deleted_at      | timestamp with time zone |           |          |
 challenge_id    | text                     |           |          |
 allocation_id   | text                     |           |          |
 blobber_id      | text                     |           |          |
 validators_id   | text                     |           |          |
 seed            | bigint                   |           |          |
 allocation_root | text                     |           |          |
 responded       | boolean                  |           |          |
Indexes:
    "challenges_pkey" PRIMARY KEY, btree (id)
    "challenge_id" UNIQUE, btree (challenge_id)
    "idx_challenges_deleted_at" btree (deleted_at)
    "idx_open_challenge" btree (blobber_id, responded, created_at)
```

File: code/go/0chain.net/smartcontract/dbs/event/challenge.go (hunk 2 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/dbs/event/challenge.go]: > this will be huge table
> as challenges are generated every round

**Review by dabasov — APPROVED** [review]
lgtm

**Comment by ssardana08** [comment]
tested
```
events_db=> EXPLAIN SELECT * FROM CHALLENGES WHERE blobber_id='h' AND responded = true AND created_at >= 111;
                                       QUERY PLAN
-----------------------------------------------------------------------------------------
 Index Scan using idx_open_challenge on challenges  (cost=0.15..8.17 rows=1 width=201)
   Index Cond: ((blobber_id = 'h'::text) AND (responded = true) AND (created_at >= 111))
```

File: code/go/0chain.net/smartcontract/storagesc/challenge.go [hunk 5]
```diff
@@ -810,6 +827,17 @@ func (sc *StorageSmartContract) populateGenerateChallenge(
 	storageChallenge.AllocationID = alloc.ID
 	storageChallenge.Created = txn.CreationDate
 
+	challInfo := &StorageChallengeInfo{
+		ID:             storageChallenge.ID,
+		Created:        storageChallenge.Created,
+		Validators:     selectedValidators,
+		RandomNumber:   seed,
+		AllocationID:   storageChallenge.AllocationID,
+		AllocationRoot: allocBlobber.AllocationRoot,
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/challenge.go]: Hmm, if the AllocationRoot is needed for challenge, why didn't we save it in StorageChallenge? I would suggest to embed StorageChallenge and add extra fields needed.

File: code/go/0chain.net/smartcontract/storagesc/challenge.go [hunk 6]
```diff
@@ -810,6 +827,17 @@ func (sc *StorageSmartContract) populateGenerateChallenge(
 	storageChallenge.AllocationID = alloc.ID
 	storageChallenge.Created = txn.CreationDate
 
+	challInfo := &StorageChallengeInfo{
+		ID:             storageChallenge.ID,
+		Created:        storageChallenge.Created,
+		Validators:     selectedValidators,
+		RandomNumber:   seed,
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/challenge.go]: Why do we need to save the seed to event db, blockchain does not use it for verification, I don't think clients need it too.

File: code/go/0chain.net/smartcontract/storagesc/challenge_eventdb.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+	"0chain.net/core/common"
+
+	"0chain.net/smartcontract/dbs"
+
+	cstate "0chain.net/chaincore/chain/state"
+	"0chain.net/smartcontract/dbs/event"
+)
+
+func storageChallengeToChallengeTable(ch *StorageChallengeInfo) *event.Challenge {
+	var validatorsArr []string
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/challenge_eventdb.go]: Init slice with capacity, rename to `validators` as Array is a different thing in Go.

File: code/go/0chain.net/smartcontract/storagesc/handler.go [hunk 8]
```diff
@@ -757,78 +757,48 @@ func (ssc *StorageSmartContract) GetValidatorHandler(ctx context.Context,
 
 }
 
-type BlobberOpenChallengesResponse struct {
-	BlobberID                string            `json:"blobber_id"`
-	ChallengeIDs             []string          `json:"challenge_ids"`
-	LatestCompletedChallenge *StorageChallenge `json:"lastest_completed_challenge"` // TODO: fix typo with Blobber and gosdk
+type StorageChallengeInfo struct {
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/handler.go]: The postfix `*Info` actually does not give much info just like `Data`  and `util`, etc. This is a struct for API response, so I suggest to use `StorageChallengeResponse` or something similar if better.

File: code/go/0chain.net/smartcontract/storagesc/handler.go [hunk 9]
```diff
@@ -757,78 +757,48 @@ func (ssc *StorageSmartContract) GetValidatorHandler(ctx context.Context,
...
+	Created        common.Timestamp  `json:"created"`
+	Validators     []*ValidationNode `json:"validators"`
+	RandomNumber   int64             `json:"seed"`
+	AllocationID   string            `json:"allocation_id"`
+	AllocationRoot string            `json:"allocation_root"`
+	BlobberID      string            `json:"blobber_id"`
+	Responded      bool              `json:"responded"`
+}
+
+type ChallengesInfo struct {
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/handler.go]: Rename to `ChallengesResponse` or something similar if better

File: code/go/0chain.net/smartcontract/storagesc/handler.go [hunk 10]
```diff
@@ -757,78 +757,48 @@ func (ssc *StorageSmartContract) GetValidatorHandler(ctx context.Context,
...
-			rsp.ChallengeIDs = make([]string, len(ocs))
-			for i, oc := range ocs {
-				rsp.ChallengeIDs[i] = oc.ID
-			}
-		}
-
-		return rsp, nil
-	default:
-		return nil, common.NewErrInternal("fail to get blobber challenge", err.Error())
+	challenges, err := getOpenChallengesForBlobber(blobberID, common.Timestamp(blobber.ChallengeCompletionTime), balances)
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/handler.go:784]: Get challenge ids of the blobber from the index of blobberChallenges first, then get the blobbers from event db without applying conditions.

**Review by peterlimg — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/smartcontract/storagesc/handler.go (hunk 10 above)
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/handler.go:780]: Is the challenges sorted by time?

**Review by peterlimg — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/smartcontract/storagesc/challenge.go (hunk 6 above)
Comment (ssardana08) [reply code/go/0chain.net/smartcontract/storagesc/challenge.go]: validator need it for verification: https://github.com/0chain/blobber/blob/c666d22325523695f77e34a29a664b9c24779849/code/go/0chain.net/validatorcore/storage/models.go#L298

File: code/go/0chain.net/smartcontract/storagesc/handler.go (hunk 10 above)
Comment (sum-it08) [reply code/go/0chain.net/smartcontract/storagesc/handler.go:780]: I have now added index sort on created_at, which will always ensure to return in sorted order

File: code/go/0chain.net/smartcontract/storagesc/challenge.go (hunk 5 above)
Comment (ssardana08) [reply code/go/0chain.net/smartcontract/storagesc/challenge.go]: done

**Review by Sriep — APPROVED** [review]

File: code/go/0chain.net/smartcontract/storagesc/handler.go (hunk 10 above)
Comment (ssardana08) [reply code/go/0chain.net/smartcontract/storagesc/handler.go:780]: fixed

File: code/go/0chain.net/smartcontract/dbs/event/challenge.go [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+
+	return &ch, nil
+}
+
+func (edb *EventDb) GetOpenChallengesForBlobber(blobberID string, now, cct common.Timestamp) ([]*Challenge, error) {
+	var chs []*Challenge
+	expiry := now - cct
+
+	result := edb.Store.Get().Model(&Challenge{}).
+		Where("created_at >= ? AND blobber_id = ? AND responded = ?",
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/dbs/event/challenge.go]: If created_at == expiry, the challenge should have expired already, so should be `created_at > ?`

File: code/go/0chain.net/smartcontract/dbs/event/challenge.go [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+	if result.Error != nil {
+		return nil, fmt.Errorf("error retriving Challenge with blobberid %v challengeid: %v; error: %v",
+			blobberID, challengeID, result.Error)
+	}
+
+	return ch, nil
+}
+
+func (edb *EventDb) addChallenge(ch *Challenge) error {
+	exists, err := ch.exists(edb)
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/dbs/event/challenge.go]: We don't need to do the existence checking, just like what we do for saving transactions to event db, if we got duplicate items, there's must be something wrong the chain, and we need to fix then.

File: code/go/0chain.net/smartcontract/dbs/event/validator.go [hunk 13]
```diff
@@ -52,6 +53,13 @@ func (edb *EventDb) GetValidatorByValidatorID(validatorID string) (Validator, er
 	return vn, nil
 }
 
+func (edb *EventDb) GetValidatorsByID(ids []string) ([]Validator, error) {
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/dbs/event/validator.go]: Perhaps rename to GetValidatorsByIDs()

File: code/go/0chain.net/smartcontract/storagesc/handler.go [hunk 14]
```diff
@@ -757,78 +757,44 @@ func (ssc *StorageSmartContract) GetValidatorHandler(ctx context.Context,
...
 }
 
-type BlobberOpenChallengesResponse struct {
-	BlobberID                string            `json:"blobber_id"`
-	ChallengeIDs             []string          `json:"challenge_ids"`
-	LatestCompletedChallenge *StorageChallenge `json:"lastest_completed_challenge"` // TODO: fix typo with Blobber and gosdk
+type StorageChallengeResponse struct {
+	*StorageChallenge `json:",inline"`
+	Validators        []*ValidationNode `json:"validators"`
+	RandomNumber      int64             `json:"seed"`
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/handler.go]: Rename RoundNumber to Seed

**Review by peterlimg — CHANGES_REQUESTED** [review]

**Review by peterlimg — APPROVED** [review]
lgtm

**Pull request merged**